Source/Decode.swift

```diff
@@ -0,0 +1,116 @@
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+
+// MARK: Decodable Parsing Methods
+
+/**
+ Parses json data at the specified key path into an object of type `T`. `T` must implement the `Decodable` protocol.
+
+ - parameter data:       A Data object containing encoded json data.
+ - parameter forKeyPath: The json key path identifying the object to be parsed. Default is `""`.
+
+ - throws:  A ParserError.Deserialization and ParserError.Validation error if parsing fails.
+ - returns: The parsed object.
+ */
+public func decodeObject<T: Decodable>(from data: Data, atKeyPath keyPath: String = "") throws -> T {
+    let entity = try Parser.parseEntity(data: data) { schema in
+        if T.self == String.self {
+            schema.addProperty(keyPath: keyPath, type: .string)
+        } else if T.self == Int.self {
+            schema.addProperty(keyPath: keyPath, type: .int)
+        } else if T.self == UInt.self {
+            schema.addProperty(keyPath: keyPath, type: .uint)
+        } else if T.self == Float.self {
+            schema.addProperty(keyPath: keyPath, type: .float)
+        } else if T.self == Double.self {
+            schema.addProperty(keyPath: keyPath, type: .double)
+        } else if T.self == Bool.self {
+            schema.addProperty(keyPath: keyPath, type: .bool)
+        } else {
+            schema.addProperty(keyPath: keyPath, type: .dictionary, decodableType: T.self)
+        }
+    }
+
+    return entity[keyPath] as! T
+}
+
+/**
+ Parses json data at the specified key path into an array of objects of type `T`. `T` must implement the
+ `Decodable` protocol.
+
+ - parameter data:       A Data object containing encoded json data.
+ - parameter forKeyPath: The json key path identifying the object to be parsed. Default is `""`.
+
+ - throws:  A ParserError.Deserialization and ParserError.Validation error if parsing fails.
+ - returns: The parsed array of objects
+ */
+public func decodeArray<T: Decodable>(from data: Data, atKeyPath keyPath: String = "") throws -> [T] {
+    let entity = try Parser.parseEntity(data: data) { schema in
+        schema.addProperty(keyPath: keyPath, type: .array, decodableType: T.self)
+    }
+
+    return entity[keyPath] as! [T]
+}
```

Comment:
Yeah, I think that works. To me it makes the `parseEntity` and `addProperty` APIs a bit more hidden if they think all the public API is in Elevate.swift, but I think it's the right call.